### ISSUE TITLE
Failed to discover network configuration error

### ISSUE THREAD

**Issue by MohanrajBabu** [convergeone/freeztp#39]
Hello,

Trying to configure cat9k switch on booting through freeZTP, but getting below failed to discover network configuration error when the switch loads and IP address not getting assigned.

```
%SYS-5-RELOAD: Reload requested by console. Reload Reason: Reload Command.
                                                                                                Chassis 1 reloading, reason - Reload command
                                                                                                                                            Mar 12 12:51:32.351 FP0/0: %PMAN-5-EXITACTION: Process manager is exiting: reload fp action requested 
Mar 12 12:51:34.79

Initializing Hardware...

System Bootstrap, Version 16.6.1r [FC1], RELEASE SOFTWARE (P)
Compiled Sat 07/15/2017  8:31:57.39 by rel

Current image running: 
Primary Rommon Image

Last reset cause: SoftwareReload
C9300-24P platform with 8388608 Kbytes of main memory

Starting network boot configuration discovery (Timeout: 25 seconds) ...
DHCP: Requesting network configuration; listening for 10 seconds
DHCP: Failed to discover network configuration (Time out)
DHCP: Backing off for 3 second(s)  
DHCP: Requesting network configuration; listening for 10 seconds
DHCP: Failed to discover network configuration (Time out)
DHCP: Backing off for 2 second(s)  
Network boot failed. Last configuration ...
 server    : 0.0.0.0 
 interface : eth0
 macaddr   : 68:66:54:34:57:80
 ip        : 0.0.0.0
 netmask   : 0.0.0.0
 gateway   : 0.0.0.0
 domain    : 
 bootfile  : 
attempting to boot from [flash:packages.conf]
```

Current setup has a centos machine connected to cat9k switch in same vlan.
ztp ip address: 20.1.1.99

ztp show config
```
#######################################################
#
#
#
ztp set suffix -confg
ztp set initialfilename network-confg
ztp set community secretcommunity
ztp set tftproot /etc/ztp/tftproot/
ztp set imagediscoveryfile freeztp_ios_upgrade
ztp set file-cache-timeout 10
ztp set snmpoid WS-C2960_SERIAL_NUMBER 1.3.6.1.2.1.47.1.1.1.1.11.1001
ztp set snmpoid WS-C3850_SERIAL_NUMBER 1.3.6.1.2.1.47.1.1.1.1.11.1000
ztp set logging merged-config-to-mainlog enable
ztp set logging merged-config-to-custom-file disable
#
#
ztp set initial-template ^
hostname {{ autohostname }}
!
snmp-server community {{ community }} RO
!
end
^
#
#
#
#######################################################
#
#
#
ztp set dhcpd INTERFACE-ETH1 subnet 20.1.1.0/24
ztp set dhcpd INTERFACE-ETH1 imagediscoveryfile-option enable
ztp set dhcpd INTERFACE-ETH1 ztp-tftp-address 20.1.1.99
ztp set dhcpd INTERFACE-ETH1 lease-time 3600
ztp set dhcpd INTERFACE-ETH1 last-address 20.1.1.210
ztp set dhcpd INTERFACE-ETH1 first-address 20.1.1.201
#
#
#
#######################################################
#
#
#
ztp set template SHORT_TEMPLATE ^
hostname {{ hostname }}
!
interface Vlan1
 ip address {{ vl1_ip_address }} 255.255.255.0
 no shut
!
end
^
#
#
#
#######################################################
#
#
#
ztp set template LONG_TEMPLATE ^
hostname {{ hostname }}
!
interface Vlan1
 ip address {{ vl1_ip_address }} {{ vl1_netmask }}
 no shut
!
!{% for interface in range(1,49) %}
interface GigabitEthernet1/0/{{interface}}
 description User Port (VLAN 1)
 switchport access vlan 1
 switchport mode access
 no shutdown
!{% endfor %}
!
ip domain-name test.com
!
username admin privilege 15 secret password123
!
aaa new-model
!
!
aaa authentication login CONSOLE local
aaa authorization console
aaa authorization exec default local if-authenticated
!
crypto key generate rsa modulus 2048
!
ip ssh version 2
!
line vty 0 15
login authentication default
transport input ssh
line console 0
login authentication CONSOLE
end
^
#
#
#
#######################################################
#
#
#
ztp set keystore DEFAULT_VALUES vl1_ip_address dhcp
ztp set keystore DEFAULT_VALUES hostname UNKNOWN_HOST
#
ztp set keystore SERIAL100 vl1_ip_address 20.1.1.201
ztp set keystore SERIAL100 hostname SOMEDEVICE
#
ztp set keystore STACK1 vl1_netmask 255.255.255.0
ztp set keystore STACK1 vl1_ip_address 20.1.1.202
ztp set keystore STACK1 hostname CORESWITCH
#
#
#
ztp set idarray STACK1 SERIAL1 SERIAL2 SERIAL3
#
#
#
ztp set association id SERIAL100 template SHORT_TEMPLATE
ztp set association id STACK1 template LONG_TEMPLATE
#
#
#
ztp set default-keystore DEFAULT_VALUES
ztp set default-template LONG_TEMPLATE
ztp set imagefile cat9k_iosxe.16.06.05.SPA.bin
ztp set image-supression 3600
ztp set delay-keystore 1000
#
#
#
#
#
#
#
#
#######################################################

```
If enabled dhcp on the cat9k interface, ip address getting assigned, but is it the right way to do it?
How to address this issue?

Thanks,
Mohan

**Comment by derek-shnosh** [comment]
I chatted with your colleague earlier and he informed me that you're running IOS-XE 16.6.3. Cisco has deprecated Smart-Install and the feature will not work on anything above 16.6.2 in that code-train.

![Screenshot_20190312_213821](https://user-images.githubusercontent.com/21090563/54255053-9bacd080-4513-11e9-915e-5ce3ee2d53b2.png)

_Reference: https://tools.cisco.com/security/center/content/CiscoSecurityAdvisory/cisco-sa-20180409-smi_

~~FreeZTP (v1.1.0) will unfortunately not be a viable option in its current form, as it uses the smart-install feature. You could still always start with an older version that still has smart-install support and use FreeZTP to upgrade to a newer version, or even use FreeZTP to generate configurations if you desired; it just couldn't be an "OOB" ZTP experience.~~

Edit: FreeZTP uses AutoInstall; as @PackeTsar mentions below, Smart-Install is not synonymous with AutoInstall... Apologies for confusion, my head was a bit unclear/distracted during our conversation.

**Comment by PackeTsar** [comment]
Hey @derek-shnosh I dont think SmartInstall being removed causes any issue with AutoInstall. I am currently deploying some 9300's running 16.10.X and FreeZTP is working fine.

**Comment by ertinostack** [comment]
@PackeTsar-Thanks so much for confirming. Would you be able to help me with the troubleshooting steps and network-config file for Cat9k. Also any specific command/feature should we turn on Cat9k switch other than (boot ipxe). Appreciate your help.

**Comment by PackeTsar** [comment]
So FreeZTP does not support iPXE for configuration of switches. FreeZTP uses the AutoInstall feature (TFTP-based) to allow the switches to download a configuration. Sorry about that.

**Comment by ertinostack** [comment]
Would you recommend any other tool that supports Cat9K iPXE  Auto provisioning

**Comment by PackeTsar** [comment]
Any particular reason you are wanting to do iPXE for the 9ks instead of AutoInstall or PnP?

**Comment by ertinostack** [comment]
Mainly for HTTP boot options since TFTP would be very slow .

**Comment by ertinostack** [comment]
we wanted to deploy close to 200 switches in 100 sites remotely..

**Comment by PackeTsar** [comment]
So TFTP is really just used to download the switch's config file (pretty much instant). Software upgrades for the newer switches is an aspect @derek-shnosh previously solved with some fancy EEM scripting, but FreeZTP can be used to get an initial config on your devices, making remote access (to manually perform software upgrades) easier.

I haven't used iPXE before, so not sure where to direct you there.

Deploying switches remotely is exactly the purpose of FreeZTP, but it's function is mainly around deploying templatized configurations to switches upon first boot. Less so around performing the software upgrades automatically.

**Comment by derek-shnosh** [comment]
Yikes! Good call @PackeTsar, I'm a bit _checked out_ mentally at the moment...

Apologies for the confusion @ertinostack.

**Comment by ertinostack** [comment]
Thanks @derek-shnosh & @PackeTsar .
@PackeTsar- Would you be able to help us with some simple Network configuration file (Cat 9K) to achieve initial config on the switches and how to software upgrades automatically.
Maybe if you can jump on webex/teamviewer for 10mins and help us would be greatly appreciated! we have testbed ready with non-working configs :-)

**Comment by ertinostack** [comment]
@PackeTsar . - Please help me with simple network-configuration file for FreeZTP and also if there is any specific config needed on Cat9k as well. 
* we just wanted to learn and achieve pushing simple network configurations to cat9k (tftp also fine)
* Upgrade new images -  TFTP also works .
Appreciate your help.!

**Comment by PackeTsar** [comment]
FreeZTP comes with a basic config by default and it will push that config to the switch upon boot. All you would have to do is set the DHCP scope options. Check out the [YouTube video](https://www.youtube.com/watch?v=x4wZvtHiyvE) for a quick tutorial.

**Comment by duewydo** [comment]
@ertinostack, did you still need help with Cat9k and freeztp? This last week I just built a FreeZTP ubuntu VM and got it working perfectly. I have tested deploying a couple cat9300 successfully at mutliple sites across the globe. I would be happy to help. We are anticipating a multi hundred stack rollout shortly and this will save my summer plans. The only thing I am torn on is the stack numbering as I don't see collecting all the switch serials and setting up the csv in advance as saving that much time. Instead since we are contracting out the grunt work we are opting to do a QC at the end and power up one at a time during cable/label/validating QC.
